src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java

```diff
@@ -121,6 +121,19 @@
     @Query("select pe from ProgrammingExercise pe where pe.assessmentType <> 'AUTOMATIC' and pe.dueDate > :#{#dateTime}")
     List<ProgrammingExercise> findAllByManualAssessmentAndDueDateAfterDate(@Param("dateTime") ZonedDateTime dateTime);
 
+    /**
+     * Returns all programming exercises that have a due date after {@code dateTime} and have tests marked with
+     * {@link de.tum.in.www1.artemis.domain.enumeration.Visibility#AFTER_DUE_DATE} but no buildAndTestStudentSubmissionsAfterDueDate.
+     * @param dateTime the time after which the due date of the exercise has to be
+     * @return List<ProgrammingExercise> (can be empty)
+     */
+    @Query("""
+                select distinct pe from ProgrammingExercise pe left join pe.testCases tc
+                where pe.dueDate > :#{#dateTime} and pe.buildAndTestStudentSubmissionsAfterDueDate is null
+                    and tc.visibility = 'AFTER_DUE_DATE'
```

Comment:
I guess here we could use uppercase to make all the sql more consistent in whole artemis.
```suggestion
                SELECT DISTINCT pe FROM ProgrammingExercise pe LEFT JOIN pe.testCases tc
                WHERE pe.dueDate > :#{#dateTime} AND pe.buildAndTestStudentSubmissionsAfterDueDate IS NULL
                    AND tc.visibility = 'AFTER_DUE_DATE'
```